### ISSUE TITLE
transforms: SAR speckle reduction filters (Lee, Refined Lee, Frost)

### ISSUE THREAD

**Issue by box1401** [torchgeo/torchgeo#3645]
## Summary

I'd like to propose adding a new submodule `torchgeo.transforms.sar` containing
classic spatial despeckling filters for SAR imagery, starting with the
**Lee (1980)** filter and extending to **Refined Lee** and **Frost** in
follow-up PRs.

torchgeo currently ships 30+ datasets that contain or are derived from
Sentinel-1 / SAR sources (`Sentinel1`, `SEN12MS`, `So2Sat`, `QuakeSet`,
`MMFlood`, `BigEarthNet-S1`, `EuroSAT-S1`, `SpaceNet6`, `ETCI2021`, etc.),
but **there are no transforms for the multiplicative speckle noise that
characterizes coherent radar imaging**. Users who want to despeckle currently
have to reach outside torchgeo (SNAP, scikit-image, custom NumPy code), which
breaks the Kornia-based on-GPU augmentation flow torchgeo relies on.

## Motivation

Speckle is the dominant noise source in SAR data. It is **multiplicative** and
non-Gaussian (typically modeled as a Gamma-distributed deviate with mean 1
and variance 1/L for L-look intensity), so generic image transforms
(Gaussian blur, AWGN augmentation) are not appropriate. The classical
adaptive filters (Lee 1980, Refined Lee 1981, Frost 1982) explicitly model
this and remain widely used as preprocessing in modern deep-learning
pipelines for SAR detection / segmentation.

A native, GPU-friendly, Kornia-compatible implementation in torchgeo would:
- Let users keep despeckling inside the PyTorch / `on_after_batch_transfer`
  pipeline.
- Make ablations (does despeckling help my detector?) trivial.
- Close the gap between SAR data ingestion and downstream models in the
  transforms module.

## Proposed API

```python
from torchgeo.transforms import LeeFilter, RefinedLeeFilter, FrostFilter

transform = LeeFilter(window_size=7, num_looks=1.0)
out = transform(sar_intensity)  # (B, C, H, W) -> same shape
```

All filters subclass `kornia.augmentation.IntensityAugmentationBase2D` so they
slot into existing `AugmentationSequential` pipelines, support `p`,
`same_on_batch`, `keepdim`, and run on GPU.

## Scope (split into 3 PRs)

| PR | Scope |
|----|-------|
| 1  | `LeeFilter` + `torchgeo.transforms.sar` module skeleton + tests + docs |
| 2  | `RefinedLeeFilter` (gradient-based edge-aware variant) |
| 3  | `FrostFilter` (exponentially weighted local mean) |

Each PR is self-contained, < 300 LOC, and individually reviewable.

## Working prototype (results)

I implemented `LeeFilter` locally to validate feasibility:

- **Numerical agreement** with a `scipy.ndimage`-based reference
  (`uniform_filter` + the standard LMMSE update): ≤ 1e-5 abs/rel difference
  on a 64×64 synthetic 1-look scene.
- **13/13 unit tests pass**: shape preservation, non-negativity, variance
  reduction in homogeneous regions, edge contrast preservation vs. plain
  box mean, and parameter validation.
- **Synthetic 1-look demo** (uniform reflectivity + bright point targets):

| Image                | ENL  |
|----------------------|------|
| Raw 1-look SAR       | 1.01 |
| Mean filter (7×7)    | 60.8 |
| **Lee filter (7×7)** | **32.9** |

The mean filter has higher ENL (more smoothing) but visibly blurs edges; Lee
preserves the bright targets and the coast-line transition (see figure below).

![Lee filter on synthetic 1-look SAR](https://github.com/user-attachments/assets/f67fbd7b-de4b-4f6b-bc6f-31535f93c617)
*Top: clean reflectivity → 1-look SAR → Lee filter (7×7) → mean filter (7×7). Bottom: horizontal profile across the bright/dark edge.*

## Open questions before opening the PR

1. **Module location** — `torchgeo/transforms/sar.py` (dedicated file for
   SAR-specific processing) seems cleanest. Would you prefer extending
   `spatial.py`?
2. **Default `num_looks`** — proposed default `1.0` (SLC). Sentinel-1 GRDH
   ships ~5-look data; happy to accept guidance on documentation here.
3. **Random vs deterministic** — `IntensityAugmentationBase2D` carries `p`
   semantics. For preprocessing `p=1.0` is the natural default; do you also
   want a plain `nn.Module` wrapper, or is `p=1.0` enough?
4. **Amplitude vs intensity** — current implementation assumes the
   multiplicative-intensity model. Should we accept amplitude/dB input via
   a flag, or document that users convert beforehand?

## Acceptance criteria for PR 1

- `torchgeo/transforms/sar.py` with `LeeFilter` class
- `torchgeo/transforms/__init__.py` updated to re-export it
- `tests/transforms/test_sar.py` with 100% coverage
- `docs/api/transforms.rst` updated with autodoc + a brief motivation
  paragraph
- Conformant to ruff / ty / prettier per `AGENTS.md`

Happy to take this on if the direction is welcome. Context: Master's thesis
work on speckle reduction for SAR ship detection.

**Comment by adamjstewart** [comment]
I think this would be great! @RitwikGupta started a PR to add many SAR-specific transforms for things like conversion from decibel-scale to power-scale, etc. but we never got around to finishing it. I think there are dozens of potential SAR transforms worth adding, and I'm happy to start with the ones you propose.

Answers to your questions:

1. `torchgeo/transforms/sar.py` seems fine with me! Could also call it `filter.py`. The exact name doesn't matter too much because it isn't part of the documented API, we can change it later if we want
2. I unfortunately don't know enough about SAR to know what would be a good default here, I would trust you and @RitwikGupta to decide
3. For now, let's go with `IntensityAugmentationBase2D` as the base class and `p=1`. @isaaccorley was talking about converting our transforms from Kornia-style to torchvision-style but I think we'll stick with Kornia for the time being
4. To keep things simple, I would assume a single input format for now. We can and should add separate transforms to convert back and forth between amplitude/power/decibel scales, but that can be done in a separate PR

There are a ton of other SAR-specific things I would love to add, including direct support for complex-valued data and complex-valued neural networks, so I see this as the beginning, not the end of improvements related to SAR support.

**Comment by box1401** [comment]
Awesome, thanks for the quick and detailed reply! @RitwikGupta happy to coordinate on direction here.

I'll start with **PR 1: `LeeFilter`** in `torchgeo/transforms/sar.py`, `IntensityAugmentationBase2D` base, `p=1` default, single-input (intensity) for now, `num_looks=1.0` default with a docstring note for multi-look (e.g. GRDH ~5-look) users. Tests + `docs/api/transforms.rst` update + ruff/ty/prettier per `AGENTS.md`.

Aiming to have a draft PR up within the week. Will ping back here when it's ready for review.

**Comment by box1401** [comment]
PR 2 of the SAR transforms series is up: #3662 — `RefinedLeeFilter`. Stacked on #3655, base set to `main` (the diff currently includes #3655's commits and will collapse on its merge — explained in the PR body).

Key design choice worth noting: it's a SNAP / GEE-style **hybrid** (refined branch when the max directional gradient exceeds ~2σ, basic-Lee box stats otherwise). The PR body has the demo numbers vs `LeeFilter` on a synthetic step edge, plus 23 new tests at 100% coverage on `torchgeo/transforms/sar.py`.

@adamjstewart no rush — happy to wait for #3655's domain review to land first if that makes the cleaner-diff route preferable.

**Comment by adamjstewart** [comment]
Yes, let's wait for #3655 before spending time on #3662, will try to encourage more people to review.
